File: Task/Abundant-deficient-and-perfect-number-classifications/Erlang/abundant-deficient-and-perfect-number-classifications-2.erl
-module(proper_divisors).
-export([classify_range/2]).

classify_range(Start, Stop) ->
    lists:foldl(fun (X, A) ->
                  Class = classify(X),
                  A#{Class => maps:get(Class, A, 0)+1} end,
                #{},
                lists:seq(Start, Stop)).

classify(N) ->
    SumPD = lists:sum(proper_divisors(N)),
    if
        SumPD  <  N -> deficient;
        SumPD =:= N -> perfect;
        SumPD  >  N -> abundant
    end.

proper_divisors(1) -> [];
proper_divisors(N) when N > 1, is_integer(N) ->
    proper_divisors(2, math:sqrt(N), N, [1]).

proper_divisors(I, L, _, A) when I > L -> lists:sort(A);
proper_divisors(I, L, N, A) when N rem I =/= 0 ->
    proper_divisors(I+1, L, N, A);
proper_divisors(I, L, N, A) when I * I =:= N ->
    proper_divisors(I+1, L, N, [I|A]);
proper_divisors(I, L, N, A) ->
    proper_divisors(I+1, L, N, [N div I, I|A]).

File: Task/Abundant-deficient-and-perfect-number-classifications/Perl/abundant-deficient-and-perfect-number-classifications-2.pl
sub div_sum {
    my($n) = @_;
    my $sum = 0;
    map { $sum += $_ unless $n % $_ } 1 .. $n-1;
    $sum;
}

my @type = <Perfect Abundant Deficient>;
say join "\n", map { sprintf "%2d %s", $_, $type[div_sum($_) <=> $_] } 1..12;
my %h;
$h{div_sum($_) <=> $_}++ for 1..20000;
say "Perfect: $h{0}    Deficient: $h{-1}    Abundant: $h{1}";
